File: grpc_api/client_python/tests/test_secured.py
import os
import sys

SOURCE_PATH = os.path.join(os.path.dirname(os.path.dirname(os.path.realpath(__file__))),"src")
sys.path.append(SOURCE_PATH)

import uuid
import dotenv
import pytest
from bbthings_grpc import Auth, Resource, DataType
from bbthings_grpc.common.utility import generate_access_key
from bbthings_grpc.common import ROOT_ID, ROOT_NAME
import utility


API_ID = uuid.uuid4()
API_PASSWORD = "Api_pa55w0rd"

PROCEDURE_ACCESS = [
    {
        "procedure": "read_model",
        "roles": ["admin", "user"]
    },
    {
        "procedure": "create_model",
        "roles": ["admin"]
    },
    {
        "procedure": "update_model",
        "roles": ["admin"]
    },
    {
        "procedure": "delete_model",
        "roles": ["admin"]
    }
]
PROCEDURES = [(uuid.uuid4(), access["procedure"]) for access in PROCEDURE_ACCESS]
ROLE_NAMES = set()
for item in PROCEDURE_ACCESS:
    for role in item["roles"]: ROLE_NAMES.add(role)
ROLES = [(uuid.uuid4(), role) for role in ROLE_NAMES]
role_access = []

ROOT_PASSWORD = "r0ot_P4s5w0rd"
ADMIN_ID = uuid.uuid4()
ADMIN_NAME = "administrator"
ADMIN_PASSWORD = "Adm1n_P4s5w0rd"
USER_ID = uuid.uuid4()
USER_NAME = "username"
USER_PASSWORD = "Us3r_P4s5w0rd"
USERS = [
    {
        "id": ADMIN_ID,
        "name": ADMIN_NAME,
        "password": ADMIN_PASSWORD,
        "role": "admin"
    },
    {
        "id": USER_ID,
        "name": USER_NAME,
        "password": USER_PASSWORD,
        "role": "user"
    }
]
user_role = []


def test_secured():
    dotenv.load_dotenv()
    address_auth = os.getenv('SERVER_ADDRESS_AUTH')
    address_resource = os.getenv('SERVER_ADDRESS_RESOURCE')
    db_auth_url_test = os.getenv("DATABASE_URL_AUTH_TEST")
    db_resource_url_test = os.getenv("DATABASE_URL_RESOURCE_TEST")

    # truncate auth and resource tables before testing
    utility.truncate_tables_auth(db_auth_url_test)
    utility.truncate_tables_resource(db_resource_url_test)

    # start auth server for testing
    utility.start_auth_server(secured=True)

    # root login
    auth_root = Auth(address_auth)
    auth_root.login(ROOT_NAME, ROOT_PASSWORD)
    assert auth_root.user_id == ROOT_ID

    # create api and procedures
    api_id = auth_root.create_api(
        id=API_ID,
        name="resource api",
        address="localhost",
        category="RESOURCE",
        description="",
        password=API_PASSWORD,
        access_key=generate_access_key()
    )
    assert api_id == API_ID
    for (proc_id, proc_name) in PROCEDURES:
        id_proc = auth_root.create_procedure(
            id=proc_id,
            api_id=api_id,
            name=proc_name,
            description=""
        )
        assert id_proc == proc_id

    # create roles and link it to procedures
    for (role_id, role_name) in ROLES:
        id_role = auth_root.create_role(
            id=role_id,
            api_id=API_ID,
            name=role_name,
            multi=False,
            ip_lock=True,
            access_duration=900,
            refresh_duration=43200
        )
        assert id_role == role_id
        for access in PROCEDURE_ACCESS:
            if any(role == role_name for role in access["roles"]):
                found = next((p for p in PROCEDURES if p[1] == access["procedure"]), None)
                procedure_id = found[0] if found else None
                add = auth_root.add_role_access(role_id, procedure_id)
                assert add == None
                role_access.append((role_id, procedure_id))
                

    # create users and link it to a role
    for user in USERS:
        user_id = auth_root.create_user(
            id=user["id"],
            name=user["name"],
            email="",
            phone="",
            password=user["password"]
        )
        assert user_id == user["id"]
        found = next((r for r in ROLES if r[1] == user["role"]), None)
        role_id = found[0] if found else None
        add = auth_root.add_user_role(user["id"], role_id)
        assert add == None
        user_role.append((user["id"], role_id))

    # test newly created admin user and regular user to login to Auth server
    auth_admin = Auth(address_auth)
    auth_admin.login(ADMIN_NAME, ADMIN_PASSWORD)
    assert auth_admin.user_id == ADMIN_ID
    auth_admin.logout()
    assert auth_admin.user_id == None
    auth_user = Auth(address_auth)
    auth_user.login(USER_NAME, USER_PASSWORD)
    assert auth_user.user_id == USER_ID
    auth_user.logout()
    assert auth_user.user_id == None

    # start resource server for testing
    utility.start_resource_server(secured=True, api_id=api_id.hex, password=API_PASSWORD)

    # admin and regular user login
    res_admin = Resource(address_resource)
    res_admin.login(address_auth, ADMIN_NAME, ADMIN_PASSWORD)
    assert res_admin.api_id() == API_ID
    assert res_admin.user_id == ADMIN_ID
    res_user = Resource(address_resource)
    res_user.login(address_auth, USER_NAME, USER_PASSWORD)
    assert res_user.api_id() == API_ID
    assert res_user.user_id == USER_ID

    # try to create model by regular user and admin user, regular user should failed and admin user should success
    with pytest.raises(Exception):
        res_user.create_model(uuid.uuid4(), "UPLINK", "name", "")
    model_id = res_admin.create_model(
        id=uuid.uuid4(),
        data_type=[DataType.F64],
        category="UPLINK",
        name="name",
        description=""
    )

    # read created model by regular user
    model = res_user.read_model(model_id)
    assert model.category == "UPLINK"
    assert model.name == "name"

    # refresh regular user
    old_token = res_user.refresh_token
    res_user.refresh()
    assert res_user.refresh_token != old_token
    # try to read model again after refreshing token
    res_user.read_model(model_id)

    # regular user and admin user logout
    res_user.logout()
    assert res_user.auth_token == None
    res_admin.logout()
    assert res_admin.auth_token == None

    # try to access resource after logout, regular and admin user should failed
    with pytest.raises(Exception):
        res_user.read_model(model_id)
    with pytest.raises(Exception):
        res_admin.read_model(model_id)

    # remove user links to role and delete user
    for (user_id, role_id) in user_role:
        auth_root.remove_user_role(user_id, role_id)
        auth_root.delete_user(user_id)

    # remove role links to procedure and delete roles
    for (role_id, proc_id) in role_access:
        auth_root.remove_role_access(role_id, proc_id)
    for (role_id, _) in ROLES:
        auth_root.delete_role(role_id)

    # delete procedures and api
    for (proc_id, _) in PROCEDURES:
        auth_root.delete_procedure(proc_id)
    auth_root.delete_api(API_ID)

    # root logout
    auth_root.logout()
    assert auth_root.auth_token == None

    # try to read api after logout, should error
    with pytest.raises(Exception):
        auth_root.read_api(API_ID)

    # stop auth and resource server
    utility.stop_auth_server()
    utility.stop_resource_server()

File: grpc_api/client_rust/tests/test_secured.rs
#[cfg(test)]
mod tests {
    use uuid::Uuid;
    use bbthings_grpc_client::{Auth, Resource, utility, DataType};
    use bbthings_grpc_server::common::test::{TestServerKind, TestServer};
    use bbthings_grpc_server::common::config::{ROOT_ID, ROOT_NAME};
    const ROOT_PASSWORD: &str = "r0ot_P4s5w0rd";

    #[tokio::test]
    async fn test_resource()
    {
        unsafe { std::env::set_var("RUST_BACKTRACE", "1"); }

        let api_id = Uuid::new_v4();
        let api_password = "Api_pa55w0rd";

        let procedure_access: &[(&str, &[&str])] = &[
            ("read_model", &["admin", "user"]),
            ("create_model", &["admin"]),
            ("update_model", &["admin"]),
            ("delete_model", &["admin"])
        ];
        let procedures: Vec<(Uuid, &str)> = procedure_access.iter()
            .map(|(procedure, _)| {
                (Uuid::new_v4(), *procedure)
            }).collect();
        let roles: Vec<(Uuid, &str)> = procedure_access.iter()
            .flat_map(|(_, roles)| *roles)
            .fold(Vec::new(), |mut acc, role| {
                if !acc.contains(&role) {
                    acc.push(role);
                }
                acc
            }).into_iter()
            .map(|r| (Uuid::new_v4(), *r))
            .collect();
        let mut role_access = Vec::new();

        let admin_id = Uuid::new_v4();
        let admin_name = "admin";
        let admin_password = "Adm1n_P4s5w0rd";
        let user_id = Uuid::new_v4();
        let user_name = "username";
        let user_password = "Us3r_P4s5w0rd";
        let users = &[
            (admin_id, admin_name, admin_password, "admin"),
            (user_id, user_name, user_password, "user")
        ];
        let mut user_role = Vec::new();

        // start auth server
        let auth_server = TestServer::new(TestServerKind::Auth);
        auth_server.truncate_tables().await.unwrap();
        auth_server.start_server();

        // root login
        let mut auth_root = Auth::new(&auth_server.address).await;
        auth_root.login(ROOT_NAME, ROOT_PASSWORD).await.unwrap();
        assert_eq!(auth_root.user_id, Some(ROOT_ID));

        // create api and procedures
        let access_key = utility::generate_access_key();
        let id = auth_root.create_api(api_id, "resource api", "localhost", "RESOURCE", "", api_password, &access_key)
            .await.unwrap();
        assert_eq!(id, api_id);
        for (proc_id, proc_name) in &procedures {
            let id = auth_root.create_procedure(*proc_id, api_id, proc_name, "")
                .await.unwrap();
            assert_eq!(id, *proc_id);
        }

        // create roles and link it to procedures
        for (role_id, role_name) in &roles {
            let id = auth_root.create_role(*role_id, api_id, role_name, false, false, 3600, 43200)
                .await.unwrap();
            assert_eq!(id, *role_id);
            for (procedure_name, role_names) in procedure_access {
                if role_names.iter().any(|r| r == role_name) {
                    let (procedure_id, _) = procedures.iter().find(|(_, p)| p == procedure_name).unwrap();
                    auth_root.add_role_access(*role_id, *procedure_id)
                        .await.unwrap();
                    role_access.push((*role_id, *procedure_id));
                }
            }
        }

        // create users and link it to a role
        for (user_id, name, password, role) in users {
            let id = auth_root.create_user(*user_id, name, "", "", password)
                .await.unwrap();
            assert_eq!(id, *user_id);
            let (role_id, _) = roles.iter().find(|(_, r)| r == role).unwrap();
            auth_root.add_user_role(*user_id, *role_id)
                .await.unwrap();
            user_role.push((*user_id, *role_id));
        }

        // test newly created admin user and regular user to login to Auth server and then logout
        let mut auth_admin = Auth::new(&auth_server.address).await;
        auth_admin.login(admin_name, admin_password).await.unwrap();
        assert_eq!(auth_admin.user_id, Some(admin_id));
        auth_admin.logout().await.unwrap();
        assert_eq!(auth_admin.user_id, None);
        let mut auth_user = Auth::new(&auth_server.address).await;
        auth_user.login(user_name, user_password).await.unwrap();
        assert_eq!(auth_user.user_id, Some(user_id));
        auth_user.logout().await.unwrap();
        assert_eq!(auth_user.user_id, None);

        // start resource server for testing
        let resource_server = TestServer::new_secured(TestServerKind::Resource, Some(&api_id.to_string()), Some(api_password));
        resource_server.start_server();

        // admin and regular user login
        let mut resource_admin = Resource::new(&resource_server.address).await;
        resource_admin.login(&auth_server.address, admin_name, admin_password).await.unwrap();
        assert_eq!(resource_admin.api_id, Some(api_id));
        assert_eq!(resource_admin.user_id, Some(admin_id));
        let mut resource_user = Resource::new(&resource_server.address).await;
        resource_user.login(&auth_server.address, user_name, user_password).await.unwrap();
        assert_eq!(resource_user.api_id, Some(api_id));
        assert_eq!(resource_user.user_id, Some(user_id));

        // try to create model by regular user and admin user, regular user should failed and admin user should success
        let try_create = resource_user.create_model(Uuid::new_v4(), &[DataType::F32T, DataType::F64T], "UPLINK", "name", None)
            .await;
        assert!(try_create.is_err());
        let model_id = resource_admin.create_model(Uuid::new_v4(), &[DataType::F32T, DataType::F64T], "UPLINK", "name", None)
            .await.unwrap();

        // read created model using user service
        let model = resource_user.read_model(model_id).await.unwrap();
        assert_eq!(model.category, "UPLINK");
        assert_eq!(model.name, "name");

        // refresh regular user and then try to read model again after refreshing token
        resource_user.refresh().await.unwrap();
        resource_user.read_model(model_id).await.unwrap();

        // regular user and admin user logout
        resource_admin.logout().await.unwrap();
        assert_eq!(resource_admin.user_id, None);
        resource_user.logout().await.unwrap();
        assert_eq!(resource_user.user_id, None);

        // try to access resource after logout, regular and admin user should failed
        let try_user = resource_user.read_model(model_id).await;
        let try_admin = resource_admin.read_model(model_id).await;
        assert!(try_user.is_err());
        assert!(try_admin.is_err());

        // remove user links to role and delete user
        for (user_id, role_id) in user_role {
            auth_root.remove_user_role(user_id, role_id).await.unwrap();
            auth_root.delete_user(user_id).await.unwrap();
        }

        // remove role links to procedure and delete roles
        for (role_id, procedure_id) in role_access {
            auth_root.remove_role_access(role_id, procedure_id).await.unwrap();
        }
        for (role_id, _) in roles {
            auth_root.delete_role(role_id).await.unwrap();
        }

        // delete procedures and api
        for (procedure_id, _) in procedures {
            auth_root.delete_procedure(procedure_id).await.unwrap();
        }
        auth_root.delete_api(api_id).await.unwrap();

        // root logout
        auth_root.logout().await.unwrap();
        assert_eq!(auth_root.user_id, None);

        // try to read api after logout, should error
        let try_read = auth_root.read_api(api_id).await;
        assert!(try_read.is_err());

        // stop auth and resource server
        resource_server.stop_server();
        auth_server.stop_server();
    }

}
